platform/linux-generic/odp_ipsec_sad.c

```diff
@@ -446,6 +446,39 @@ static uint32_t esp_block_len_to_mask(uint32_t block_len)
 	return block_len - 1;
 }
 
+/* AR window management initialization */
+static int ipsec_antireplay_init(ipsec_sa_t *ipsec_sa,
+				 const odp_ipsec_sa_param_t *param)
+{
+	uint16_t num_bkts = 0;
+
+	if (param->inbound.antireplay_ws > IPSEC_AR_WIN_SIZE_MAX) {
+		ODP_ERR("AR_WS[%u] not supported\n", param->inbound.antireplay_ws);
+		return -1;
+	}
+
+	ipsec_sa->antireplay = (param->inbound.antireplay_ws != 0);
+	if (!ipsec_sa->antireplay)
+		return 0;
+
+	ipsec_sa->hot.in.win_size = param->inbound.antireplay_ws;
+	/* Window size should be at least IPSEC_AR_WIN_SIZE_MIN */
+	if (ipsec_sa->hot.in.win_size < IPSEC_AR_WIN_SIZE_MIN)
+		ipsec_sa->hot.in.win_size = IPSEC_AR_WIN_SIZE_MIN;
+
+	num_bkts = ROUNDUP_POWER2_U32(ipsec_sa->hot.in.win_size) / IPSEC_AR_WIN_BUCKET_SIZE;
```

Comment:
This is broken in a way that makes it possible for replayed packets to pass the anti-replay check.

There should be at least one extra bucket so that when the window is not aligned with the buckets, the top and bottom bucket do not end up being the same. Now they do, and thus the clearing of the new top bucket also clears the bits that are still used for the bottom of the window.